experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,2237 @@
+//===----------------------------------------------------------------------===//
+/// Arithmetic and generic components
+
+class TruncIModel : public OpExecutionModel<handshake::TruncIOp> {
+public:
+  using OpExecutionModel<handshake::TruncIOp>::OpExecutionModel;
+  TruncIModel(handshake::TruncIOp trunciOp, mlir::DenseMap<Value, RW *> &subset)
+      : OpExecutionModel<handshake::TruncIOp>(trunciOp), op(trunciOp),
+        inputWidth(op.getIn().getType().getDataBitWidth()),
+        outputWidth(op.getResult().getType().getDataBitWidth()) {
+    ins = getState<ChannelConsumerRW>(op.getOperand(), subset);
+    outs = getState<ChannelProducerRW>(op.getResult(), subset);
+  }
+
+  void reset() override {
+    const APInt k = dataCast<const APInt>(ins->data);
+    outs->data = k.trunc(outputWidth);
+    outs->valid = ins->valid;
+    ins->ready = !ins->valid || (ins->valid && outs->ready);
+  }
+
+  void exec(bool isClkRisingEdge) override { reset(); }
+
+  void printStates() override {
+    printValue<ConsumerRW, const Data>("ins", ins, &ins->data);
+    printValue<ProducerRW, Data>("outs", outs, &outs->data);
+  }
+
+private:
+  handshake::TruncIOp op;
+
+  // parameters
+  unsigned inputWidth;
+  unsigned outputWidth;
+
+  // ports
+  ChannelConsumerRW *ins;
+  ChannelProducerRW *outs;
+};
+
+using UnaryCompFunc = std::function<Data(const Data &, unsigned)>;
+
+/// Class to represent NEGF, NOT, EXTSI, EXTUI
+template <typename Op>
+class GenericUnaryOpModel : public OpExecutionModel<Op> {
+public:
+  using OpExecutionModel<Op>::OpExecutionModel;
+  GenericUnaryOpModel(Op op, mlir::DenseMap<Value, RW *> &subset,
+                      const UnaryCompFunc &callback)
+      : OpExecutionModel<Op>(op), op(op),
+        outputWidth(op.getResult().getType().getDataBitWidth()),
+        callback(callback) {
+    ins = OpExecutionModel<Op>::template getState<ChannelConsumerRW>(
+        op.getOperand(), subset);
+    outs = OpExecutionModel<Op>::template getState<ChannelProducerRW>(
+        op.getResult(), subset);
+  }
+
+  void reset() override {
+    outs->data = callback(ins->data, outputWidth);
+    outs->valid = ins->valid;
+    ins->ready = outs->ready;
+  }
+
+  void exec(bool isClkRisingEdge) override { reset(); }
+
+  void printStates() override {
+    OpExecutionModel<Op>::template printValue<ConsumerRW, const Data>(
+        "ins", ins, &ins->data);
+    OpExecutionModel<Op>::template printValue<ProducerRW, Data>("outs", outs,
+                                                                &outs->data);
+  }
+
+private:
+  Op op;
+
+  // parameters
+  unsigned outputWidth;
+  UnaryCompFunc callback;
+
+  // ports
+  ChannelConsumerRW *ins;
+  ChannelProducerRW *outs;
+};
+
+/// Mutual component for binary operations: AddI, AndI, CmpI, MulI, OrI,
+/// ShlI, ShrsI, ShruI, SubI, XorI...
+using BinaryCompFunc = std::function<Data(const Data &, const Data &)>;
+template <typename Op>
+class GenericBinaryOpModel : public OpExecutionModel<Op> {
```

Comment:
Modulo my comment in exec this seems nice! People will be able to use this easily for all kinds of operations, so it's super useful.